### PR TITLE
Fix standard deviation overflow(#95)

### LINKED ISSUE #95: Standard deviation value is wrong (overflow may happen)
Now I'm testing #94, I find this issue.

## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 18.04 (64 bit)
- Installation type:
  - build Eloquent
  - master also has this problem
- DDS implementation:
  - unrelated
- Client library (if applicable):
  - rttest

#### Steps to reproduce issue
- run example_loop, and we get following stdout.

```
# ./example_loop -d 1gb -s rr -i 10000 -f out.txt
Writing results to file: out.txt
Initial major pagefaults: 0
Initial minor pagefaults: 262500
rttest statistics for new.txt:
  - Minor pagefaults: 0
  - Major pagefaults: 0
  Latency (time after deadline was missed):
    - Min: 9792 ns
    - Max: 570881 ns
    - Mean: 82055.158300 ns
    - Standard deviation: 141.661569
```

- If you run many times, you'll get large standard deviation such as 42949672.959047.

- Anyway, run following to calcurate true standard deviation.
  Different values are returned.

```
# run in ipython session
import numpy as np
np.loadtxt("out.txt", skiprows=1)[:, 2].std()
# 10,000 order value is returned, for example 27376.849010133381
```

- I think overflow happens.

#### Expected behavior
see above

#### Actual behavior
see above

#### Additional information
To calcurate standard deviation, Rttest::calculate_statistics does following (I wrote python-numpy like formula after "i.e.".)
- (1) minus mean i.e. `x = x - x.mean()` where x is jitter vector.
- (2) get inner product i.e. `x = np.dot(x, x)`
- (3) divide by N and sqrt i.e. `x = math.sqrt(x/N)`

To reduce the possibility of overflow, I propose to divide by sqrt(N) first.
- (1) minus mean i.e. `x = x - x.mean()`
- (2) divide by sqrt(N) i.e. `x = x / sqrt(N)`
- (3) get inner product i.e. `x = np.dot(x, x)`
- (4) we already divide, so only take sqrt i.e. `x = math.sqrt(x)`

Here is my workaround code(I get almost same output as numpy).
If OK, I'll send PR.

```
// Rttest::calculate_statistics in rttest.cpp
// https://github.com/ros2/realtime_support/blob/master/rttest/src/rttest.cpp#L839

// before
  std::vector<int64_t> latency_diff(latency_dataset.size());
  std::transform(
    latency_dataset.begin(), latency_dataset.end(), latency_diff.begin(),
    std::bind(std::minus<int>(), std::placeholders::_1, output->mean_latency));
  int64_t sq_sum = std::inner_product(
    latency_diff.begin(), latency_diff.end(), latency_diff.begin(), 0);
  output->latency_stddev = std::sqrt(sq_sum / latency_dataset.size());


// after
   std::transform(
     latency_dataset.begin(), latency_dataset.end(), latency_diff.begin(),
     std::bind(std::minus<int>(), std::placeholders::_1, output->mean_latency));
  // forst divide by sqrt(n)
  auto n = latency_dataset.size();
  std::vector<double> latency_div(n);
  std::transform(
	  latency_diff.begin(), latency_diff.end(), latency_div.begin(),
	  [n](int64_t x) -> double { return x / std::sqrt(n); });
  // inner_product
   int64_t sq_sum = std::inner_product(
     latency_div.begin(), latency_div.end(), latency_div.begin(), 0);
   output->latency_stddev = std::sqrt(sq_sum);
```

### PR DESCRIPTION
**Pull request by y-okumura-isp**
Since I haven't written unit test, l paste the example_loop and numpy results instead.
I checked in 64 bit(ubuntu x86_64) and 32 bit(Raspberry Pi B+), pasted results are in 32bit.

Standard deviations are not equal exactly because it's a floating point calculation.

```
# ./example_loop -s rr -i 10000 -d 200mb -f out.txt
Writing results to file: out.txt
Initial major pagefaults: 0
Initial minor pagefaults: 51508
rttest statistics for out.txt:
  - Minor pagefaults: 0
  - Major pagefaults: 0
  Latency (time after deadline was missed):
    - Min: 8302 ns
    - Max: 135160 ns
    - Mean: 11098.962400 ns
    - Standard deviation: 2934.104804

# python
>>> import numpy as np
>>> np.loadtxt("out.txt", skiprows=1)[:, 2].std()
2934.929376524457
```

### CODE REVIEW

**Comment by dirk-thomas** [comment]
@y-okumura-isp please update the PR title and use a descriptive text of the actual change. The reference to the issue this addresses should then go into the description with a proper keyword so GitHub closes the referenced issue automatically when this PR is merged.

**Comment by y-okumura-isp** [comment]
@clalancette Thank you for comment.

> Please add a test when you have some time, that way we can ensure we don't break it in the future.

I did in 3be4434.
There are many changes, so let me explain.

(1) math_utils.hpp, test_math_utils.cpp
  - As calculation routine is embedded in `Rttest::calculate_statistics`, so I needed to split this as a function and write test.
    As utils.h is C-style, I added math_utils.hpp (although my PR 96 changes utils.h to utils.hpp...) 
  - test_api.cpp seems to be test for rttest API, so I added test_math_utils.cpp.

(2) minnor fix
  - Type of `latency_diff` was `std::vector<int64_t>`, but `std::vector<double>` is preferable because average (not integer) is subtracted.

By the way as calcurating statistics is very common, I think it's better to use some statistics library or math library such as eigen.
But I didn't because changes will be bigger.

**Comment by clalancette** [comment]
> By the way as calcurating statistics is very common, I think it's better to use some statistics library or math library such as eigen.
> But I didn't because changes will be bigger.

I agree with not bringing in another dependency.

File: rttest/include/rttest/math_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
```
Comment (clalancette) [inline rttest/include/rttest/math_utils.hpp:1]:
```suggestion
// Copyright 2020 Open Source Robotics Foundation, Inc.
```

File: rttest/include/rttest/math_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RTTEST__MATH_UTILS_HPP_
+#define RTTEST__MATH_UTILS_HPP_
+
+#include <vector>
+#include <cmath>
+#include <algorithm>
+#include <numeric>
```
Comment (clalancette) [inline rttest/include/rttest/math_utils.hpp:21]: Nit: alphabetize this list.

File: rttest/include/rttest/math_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  double mean = std::accumulate(
+    vec.begin(),
+    vec.end(), 0.0) / n;
+
+  std::vector<double> diff(n);
+  std::transform(
+    vec.begin(), vec.end(), diff.begin(),
+    [mean](auto x) -> double {return x - mean;});
+
+  // forst divide by sqrt(n)
```
Comment (clalancette) [inline rttest/include/rttest/math_utils.hpp:36]:
```suggestion
  // first divide by sqrt(n)
```

File: rttest/test/test_math_util.cpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+#include "gtest/gtest.h"
+#include "rttest/math_utils.hpp"
+
+TEST(MathUtils, calculate_stddev) {
+  std::vector<uint64_t> v_int = {1, 2, 3, 4};
+  EXPECT_NEAR(1.118, calculate_stddev(v_int), 0.001);
+
+  std::vector<double> v_double = {1.1, 2.2, 3.3, 4.9};
+  EXPECT_NEAR(1.404, calculate_stddev(v_double), 0.001);
+}
```
Comment (clalancette) [inline rttest/test/test_math_util.cpp:26]: Can you please add a test here that shows the problem from #95?

File: rttest/include/rttest/math_utils.hpp (hunk 2 above)
Comment (y-okumura-isp) [reply rttest/include/rttest/math_utils.hpp]: 6993c6f

File: rttest/test/test_math_util.cpp (hunk 4 above)
Comment (y-okumura-isp) [reply rttest/test/test_math_util.cpp:46]: bcf95fe.

**Comment by y-okumura-isp** [comment]
Thank you for many comments.
I've done with all comments(if no oversight).

**Comment by clalancette** [comment]
Thanks for the contribution; this looks good to me now.  I'm going to hold off on running CI and merging this until after Foxy is released; please feel free to ping me after June 8th if I forget about this.

**Comment by clalancette** [comment]
@y-okumura-isp Oh, one other thing; can you please rebase and add a "Signed-off-by" line to all of your commits?  That way our DCO check will pass.  Thanks.

**Comment by y-okumura-isp** [comment]
> can you please rebase and add a "Signed-off-by" line to all of your commits?

Sure, I fixed.
Thank you for reviewing so patiently.

**Comment by clalancette** [comment]
> Thank you for reviewing so patiently.

You are welcome, thanks for iterating with us.

**Comment by clalancette** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11038/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6363/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8988/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10959/)

**Comment by clalancette** [comment]
Windows is failing CI, but that's because of infrastructure issues.  Since this package doesn't build on Windows *anyway*, I'm going to go ahead and merge without that.  Thanks for the contribution!

**Pull request merged**